src/main/scala/com/fulcrumgenomics/bam/ComputeRnaSeqInsertSize.scala

```diff
@@ -0,0 +1,330 @@
+    else if (rec.getSecondOfPairFlag) return None
+
+    val mateCigar        = SAMUtils.getMateCigar(rec)
+    require(mateCigar != null, s"Mate CIGAR (Tag MC) not found for $rec.")
+    val mateAlignmentEnd = CoordMath.getEnd(rec.getMateAlignmentStart, mateCigar.getReferenceLength)
+
+    // Check for overlapping reads
+    if (CoordMath.overlaps(rec.getAlignmentStart, rec.getAlignmentEnd, rec.getMateAlignmentStart, mateAlignmentEnd)) {
+      fail(s"Record overlaps with its mate, consider using ClipOverlappingReads.  Record: $rec")
+    }
+
+    // Get the overlapping genes
+    val leftMostBase     = Math.min(rec.getAlignmentStart, rec.getMateAlignmentStart)
+    val rightMostBase    = Math.max(rec.getAlignmentEnd, mateAlignmentEnd)
+    val recInterval      = new Interval(rec.getReferenceName, leftMostBase, rightMostBase)
+    val overlappingGenes = overlapDetector.getOverlaps(recInterval)
+
+    if (overlappingGenes.size() == 1) {
+      val mateAlignmentBlocks = SAMUtils.getAlignmentBlocks(mateCigar, rec.getMateAlignmentStart, "mate cigar")
+      val (lowerReadAlignmentBlocks, upperReadAlignmentBlocks) = if (rec.getAlignmentStart < rec.getAlignmentEnd) {
+        (rec.getAlignmentBlocks, mateAlignmentBlocks)
+      }
+      else {
+        (mateAlignmentBlocks, rec.getAlignmentBlocks)
+      }
+
+      val gene = overlappingGenes.head
+      calculateInsertSizeFromRemainingTranscripts(
+        gene                     = gene,
+        lowerReadAlignmentBlocks = lowerReadAlignmentBlocks,
+        upperReadAlignmentBlocks = upperReadAlignmentBlocks,
+        remainingTranscripts     = gene.toIterator
+      )
+    }
+    else {
+      None
+    }
+  }
+
+  /**
+    * Calculates the insert size for all remaining transcripts ensuring that the results are the same.  If a transcript
+    * is encountered that yields an insert size different than previously calculated, computation is stopped, and None
+    * is returned.
+    */
+  @tailrec
+  private def calculateInsertSizeFromRemainingTranscripts(gene: Gene,
+                                                          lowerReadAlignmentBlocks: Iterable[AlignmentBlock],
+                                                          upperReadAlignmentBlocks: Iterable[AlignmentBlock],
+                                                          remainingTranscripts: Iterator[Transcript],
+                                                          result: Option[Int] = None): Option[Int] = {
+    if (remainingTranscripts.isEmpty) { // no more transcripts, return the result
+      result
+    }
+    else {
+      // calculate the result for this transcript
+      val transcript = remainingTranscripts.next()
+      val exons      = transcript.toIterator.bufferBetter
+
+      // calculate the # of bases overlapping the lower read
+      val lower = numReadBasesOverlappingTranscript(lowerReadAlignmentBlocks.toIterator.bufferBetter, exons)
```

Comment:
I was thinking this whole calculation could be a little simpler.  My thinking is that you could do:
1) Take the 5' end position of both reads in the pair
2) Calculate the transcript coordinate of the two 5' ends
3) Calculate the length in transcript space as the distance between the two ends on the transcript

Admittedly that doesn't account for indels in the reads themselves, but generally speaking other insert size tools don't either, I think.